R/summary-log.R

```diff
@@ -266,3 +281,153 @@ do_if_bbi_sum <- function(fn, mode) {
     return(res)
   }
 }
+
+#' Calculate Akaike Information Criterion (AIC) and Bayesian Information
+#' Criterion (BIC), and append to summary log
+#' @param .log_df The output of [summary_log()]) or `run_log() %>% add_summary()`
+#' @keywords internal
+add_aic_bic <- function(.log_df){
+  .log_df %>% dplyr::mutate(
+    !!rlang::sym(AIC_COL) := 2*.data[[PARAM_COUNT_COL]] + .data[[OFV_COL]],
+    !!rlang::sym(BIC_COL) := .data[[PARAM_COUNT_COL]]*log(.data$number_of_obs) + .data[[OFV_COL]]
+  ) %>% dplyr::relocate(all_of(c(AIC_COL, BIC_COL)), .after = all_of(OFV_COL))
+}
+
+
+#' Calculate the difference in objective function value
+#'
+#' Calculate the difference in objective function value (`ofv`) between the
+#' `.bbi_object` and a reference model.
+#'
+#' @details
+#' If `.bbi_object` is a `tibble` of class `bbi_log_df`, `"absolute_model_path"`
+#' is the only required column. If the `tibble` is the output of [bbr::run_log()]
+#' (i.e. no summary information is appended), the `"ofv"` column will appended
+#' as well for added context.
+#'
+#'
+#' @param .bbi_object The object to compare. Can be a `bbi_nonmem_model` object
+#'  or a tibble of class `bbi_log_df`.
+#' @param .mod2 If a `bbi_nonmem_model` object is passed, calculate the
+#'  difference in objective function value between each model in `.bbi_object`
+#'  and `.mod2`. If `.mod2 = NULL`, the default, calculate the difference
+#'  between each model in `.bbi_object` and the model at `get_based_on(.bbi_object)`.
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # Calculate for individual model
+#' add_dofv(mod3, .mod2 = NULL) # uses based_on model
+#' add_dofv(mod3, .mod2 = mod2) # provide reference model
+#'
+#' # Append to run logs:
+#' run_log(MODEL_DIR) %>% add_summary() %>% add_dofv()
+#' summary_log(MODEL_DIR) %>% add_dofv()
+#'
+#' # Automatically adds `ofv` column in absence of `add_summary()` call:
+#' run_log(MODEL_DIR) %>% add_dofv()
+#' }
+#'
+#'
+#' @seealso [add_summary()]
+#' @return
+#' If `.bbi_object` is a `tibble`, `add_dofv()` appends a `dofv` column.
+#'
+#' If `.bbi_object` is a `bbi_nonmem_model` object, `add_dofv()` returns a numeric value.
+#' @export
+add_dofv <- function(.bbi_object, .mod2 = NULL){
+  UseMethod("add_dofv")
+}
+
+#' @export
+add_dofv.bbi_nonmem_model <- function(.bbi_object, .mod2 = NULL){
```

Comment:
`add_dofv` doesn't seem like a good name for a method that takes a model and returns an scalar (rather than adding/extending to the passed log).

My opinion is that, at least at this point, we shouldn't bother adding an S3 method and should instead convert `add_dofv` to a regular function that takes a run/summary log.

Another option would be to change the name to something that fits both scenarios better.

(I realize there was some previous discussion that landed on this. I discussed this offline with @seth127 and he was on board with switching to a regular function. He can of course chime in if his thoughts have since changed on that.)